File: music-universe/music-data-raw-lastfm/src/main/java/yurykorzun/art/universe/music/data/raw/lastfm/collectable/tag/entity/LastfmTag.java
package yurykorzun.art.universe.music.data.raw.lastfm.collectable.tag.entity;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.SuperBuilder;

import jakarta.persistence.*;
import yurykorzun.art.universe.music.data.raw.lastfm.collectable.common.entity.BaseLastfmEntity;
import yurykorzun.art.universe.music.data.raw.lastfm.collectable.common.entity.LastfmEntityType;
import yurykorzun.art.universe.music.data.raw.lastfm.common.LastfmConstants;

@Entity(name = "tag")
@SuperBuilder
@NoArgsConstructor
@Getter
public class LastfmTag extends BaseLastfmEntity {

    @Id
    @SequenceGenerator(
            name = "tag_seq_gen",
            sequenceName = "tag_seq",
            allocationSize = LastfmConstants.HIBERNATE_BATCH_SIZE
    )
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "tag_seq_gen")
    private long id;

    @Setter
    @Column(name = "usage_count")
    private int usageCount;

    @Setter
    @Column(name = "usage_users_count")
    private int usageUsersCount;

    @Override
    @Transient
    public LastfmEntityType getType() {
        return LastfmEntityType.TAG;
    }

    @Override
    public String getUniqueKey() {
        return getName();
    }
}

File: music/ui/src/music/data/raw/lastfm/types/lastfm-tag.ts
import {BaseLastfmEntity} from "./lastfm-base-entity.ts";
import type {Category} from "@/music/shared/types/entities.ts";
import type {ApprovalStatusType} from "@/music/data/raw/lastfm/constants/approvalStatus.ts";

export class LastfmTag extends BaseLastfmEntity<"category"> {
    url: string | null;
    usageCount: number | null;
    usageUsersCount: number | null;

    constructor(
        id: number,
        name: string,
        url: string | null,
        approvalStatus: ApprovalStatusType,
        usageCount: number | null,
        usageUsersCount: number | null,
        masterEntity?: Category
    ) {
        super(id, name, approvalStatus, masterEntity);
        this.url = url;
        this.usageCount = usageCount;
        this.usageUsersCount = usageUsersCount;
    }

    getEntityType(): "category" {
        return "category";
    }
}
